### Tidy Data Exercises

**1.  Attendance Data**

**Load the ```attendance.csv``` file and calculate an attendance percentage for each student. One half day is worth 50% of a full day, and 10 tardies is equal to one absence.**

**You should end up with something like this:**

**```
name
Billy    0.5250
Jane     0.6875
John     0.9125
Sally    0.7625
Name: grade, dtype: float64```**

In [52]:
import pandas as pd

attendance = pd.read_csv('attendance.csv')
attendance

attendance.rename(columns={'Unnamed: 0':'name'}, inplace=True)
attendance

name 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05 2018-01-06  \
0  Sally          P          T          T          H          P          A   
1   Jane          A          P          T          T          T          T   
2  Billy          A          T          A          A          H          T   
3   John          P          T          H          P          P          T   

  2018-01-07 2018-01-08  
0          T          T  
1          A          T  
2          P          T  
3          P          P

In [53]:
# restructure data using 'melt'
attendance = attendance.melt(id_vars=['name'], var_name='days', value_name='score')
attendance


name        days score
0   Sally  2018-01-01     P
1    Jane  2018-01-01     A
2   Billy  2018-01-01     A
3    John  2018-01-01     P
4   Sally  2018-01-02     T
5    Jane  2018-01-02     P
6   Billy  2018-01-02     T
7    John  2018-01-02     T
8   Sally  2018-01-03     T
9    Jane  2018-01-03     T
10  Billy  2018-01-03     A
11   John  2018-01-03     H
12  Sally  2018-01-04     H
13   Jane  2018-01-04     T
14  Billy  2018-01-04     A
15   John  2018-01-04     P
16  Sally  2018-01-05     P
17   Jane  2018-01-05     T
18  Billy  2018-01-05     H
19   John  2018-01-05     P
20  Sally  2018-01-06     A
21   Jane  2018-01-06     T
22  Billy  2018-01-06     T
23   John  2018-01-06     T
24  Sally  2018-01-07     T
25   Jane  2018-01-07     A
26  Billy  2018-01-07     P
27   John  2018-01-07     P
28  Sally  2018-01-08     T
29   Jane  2018-01-08     T
30  Billy  2018-01-08     T
31   John  2018-01-08     P

In [49]:
attendance.score.replace('P',1).replace('A',0).replace('T',.9).replace('H',.5)


0     1.0
1     0.0
2     0.0
3     1.0
4     0.9
5     1.0
6     0.9
7     0.9
8     0.9
9     0.9
10    0.0
11    0.5
12    0.5
13    0.9
14    0.0
15    1.0
16    1.0
17    0.9
18    0.5
19    1.0
20    0.0
21    0.9
22    0.9
23    0.9
24    0.9
25    0.0
26    1.0
27    1.0
28    0.9
29    0.9
30    0.9
31    1.0
Name: score, dtype: float64

**2. Coffee Levels**

**a. Read the ```coffee_levels.csv``` file.**


In [69]:
coffee = pd.read_csv('coffee_levels.csv')
coffee.head()


coffee.rename(columns={'coffee_carafe':'carafe'}, inplace=True)
coffee

hour carafe  coffee_amount
0      8      x       0.816164
1      9      x       0.451018
2     10      x       0.843279
3     11      x       0.335533
4     12      x       0.898291
5     13      x       0.310711
6     14      x       0.507288
7     15      x       0.215043
8     16      x       0.183891
9     17      x       0.391560
10     8      y       0.189297
11     9      y       0.521502
12    10      y       0.023163
13    11      y       0.235529
14    12      y       0.017009
15    13      y       0.997464
16    14      y       0.058361
17    15      y       0.144644
18    16      y       0.544676
19    17      y       0.594126
20     8      z       0.999264
21     9      z       0.915990
22    10      z       0.144928
23    11      z       0.311495
24    12      z       0.771947
25    13      z       0.398520
26    14      z       0.864464
27    15      z       0.436364
28    16      z       0.280621
29    17      z       0.436677

**b. Transform the data so that each carafe is in it's own column.**


In [77]:
tidy_coffee = coffee.pivot(index = 'hour', columns = 'carafe')
tidy_coffee

coffee_amount                    
carafe             x         y         z
hour                                    
8           0.816164  0.189297  0.999264
9           0.451018  0.521502  0.915990
10          0.843279  0.023163  0.144928
11          0.335533  0.235529  0.311495
12          0.898291  0.017009  0.771947
13          0.310711  0.997464  0.398520
14          0.507288  0.058361  0.864464
15          0.215043  0.144644  0.436364
16          0.183891  0.544676  0.280621
17          0.391560  0.594126  0.436677

In [78]:

tidy_coffee = tidy_coffee.set_index(['hour', 'carafe']).unstack(level=1)
#coffee.columns.droplevel()
tidy_coffee

KeyError: "None of ['hour', 'carafe'] are in the columns"

**c. Is this the best shape for the data?**

**3.  Cake Recipes**

**a. Read the ```cake_recipes.csv``` data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures.**


In [79]:
recipes = pd.read_csv('cake_recipes.csv')
recipes


recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084
5           c:top  71.306308  82.795477  92.098049  53.960273
6        d:bottom  52.799753  58.670419  51.747686  56.183110
7           d:top  96.873178  76.101363  59.571620  50.971626

**b. Tidy the data as necessary.**


**c. Which recipe, on average, is the best? recipe b


**d. Which oven temperature, on average, produces the best results?** 275


**e. Which combination of recipe, rack position, and temperature gives the best result?** recipe b, bottom rack, 300 degrees